src/Payments.sol

```diff
@@ -39,6 +39,19 @@ contract Payments is
 {
     using SafeERC20 for IERC20;
     using RateChangeQueue for RateChangeQueue.Queue;
+    
+    string public constant VERSION = "0.0.0";
+    
+    event ContractUpgraded(string version, address implementation);
+    
+    function getVersion() public pure returns (string memory) {
+        return VERSION;
+    }
+    
+    function postUpgrade() external {
+        require(msg.sender == address(this), "Only callable via delegatecall from proxy");
+        emit ContractUpgraded(VERSION, address(this));
```

Comment:
Looking into this a big more this won't be the expected implementation value as written:   https://ethereum.stackexchange.com/questions/90201/what-is-the-value-of-addressthis-when-delegate-calling

address(this) is the proxy not the implementation.

Reply:
We'll want to grab this from the IMPLEMENTATIN_SLOT

Reply:
This one can be copied over here: https://github.com/FilOzone/pdp/pull/160/files